modules/standard/Spawn.chpl

```diff
@@ -779,82 +810,99 @@ module Spawn {
-                when waiting for the child process.
    */
-  proc subprocess.communicate(out error:syserr) {
+  proc subprocess.communicate() throws {
+    try _throw_on_launch_error();
 
+    var err:syserr = ENOERR;
     on home {
       if this.stdin_pipe {
         // send data to stdin
         _stop_stdin_buffering();
       }
 
-      error = qio_proc_communicate(
+      err = qio_proc_communicate(
         locking,
         stdin_channel._channel_internal,
         stdout_channel._channel_internal,
         stderr_channel._channel_internal);
     }
+    if err then try ioerror(err, "in subprocess.communicate");
 
-    if !error {
-      // wait for child process to terminate
-      var done:c_int = 0;
-      var exitcode:c_int = 0;
-      error = qio_waitpid(pid, 1, done, exitcode);
-      if done {
-        this.running = false;
-        this.exit_status = exitcode;
-      }
+    // wait for child process to terminate
+    var done:c_int = 0;
+    var exitcode:c_int = 0;
+    err = qio_waitpid(pid, 1, done, exitcode);
+    if done {
+      this.running = false;
+      this.exit_status = exitcode;
     }
+    if err then try ioerror(err, "in subprocess.communicate");
   }
 
-  // documented in the out error version
+  // documented in the throws version
   pragma "no doc"
-  proc subprocess.communicate() throws {
-    var err:syserr = ENOERR;
-
-    try _throw_on_launch_error();
-
-    this.communicate(error=err);
-    if err then try ioerror(err, "in subprocess.communicate");
+  proc subprocess.communicate(out error:syserr) {
+    compilerWarning("'out error: syserr' pattern has been deprecated, use 'throws' function instead");
+    try {
+      this.communicate();
+    } catch e: SystemError {
+      error = e.err;
+    } catch {
+      error = EINVAL;
+    }
   }
 
   /*
     Close any open channels and pipes for interacting with a subprocess.  This
     function does not wait for the subprocess to complete.  Note that it is
     generally not necessary to call this function since these channels will be
     closed when the subprocess record goes out of scope.
-
-    :arg error: optional argument to capture any error encountered
-                when closing the pipes.
    */
-  proc subprocess.close(out error:syserr) {
-    error = ENOERR;
+  proc subprocess.close() throws {
+    // TODO: see subprocess.wait() for more on this error handling approach
+    var err: syserr = ENOERR;
+
     // Close stdin.
     if this.stdin_pipe {
       // send data to stdin
       _stop_stdin_buffering();
-      this.stdin_channel.close(error=error);
+      try {
+        this.stdin_channel.close();
+      } catch e: SystemError {
+        err = e.err;
+      } catch { }
     }
     // Close stdout.
     if this.stdout_pipe {
-      this.stdout_channel.close(error=error);
+      try {
+        this.stdout_channel.close();
+      } catch e: SystemError {
+        err = e.err;
+      } catch { }
```

Comment:
Same issues as the other one, don't want empty catchall, should be using different error variables for the different calls and combining them at the end.